File: web/src/lib/accent.ts
// The design system's accent presets — the brand hue the primary/ring tokens paint.
//
// Mirrors `api/src/accent.rs` (`SUPPORTED_ACCENTS`), like `lifeLayout.ts` mirrors the
// layout slugs: a slug added on one side only is either rejected by the API or renders
// as the default. The colour values themselves live in `src/assets/main.css` as
// `[data-accent='<slug>']` token blocks (every preset's light/dark pairs are
// WCAG-AA-checked — see docs/design-system.md), which is why only preset slugs exist:
// a free-form colour could not carry those guarantees. `swatch` is the preset's vivid
// (dark-mode primary) hex, used only to paint the settings picker dots.

export const ACCENT_OPTIONS = [
  { value: 'pink', label: 'Pink', swatch: '#f471a8' },
  { value: 'ember', label: 'Ember', swatch: '#f28e42' },
  { value: 'violet', label: 'Violet', swatch: '#ab91f2' },
  { value: 'teal', label: 'Teal', swatch: '#39bab4' },
  { value: 'blue', label: 'Blue', swatch: '#58a5e4' },
  { value: 'green', label: 'Green', swatch: '#5ebc7b' },
] as const

export type Accent = (typeof ACCENT_OPTIONS)[number]['value']

export const DEFAULT_ACCENT: Accent = 'pink'

export function isAccent(value: unknown): value is Accent {
  return ACCENT_OPTIONS.some((option) => option.value === value)
}

File: api/src/accent.rs
//! Supported UI accent colours — the account's brand-hue preference.
//!
//! The SPA's design system paints its primary/ring tokens from a small set of
//! AA-validated presets (see `docs/design-system.md`); the chosen slug lives on the
//! account so the accent follows a user across devices. The server stores and validates
//! the *slug only* — the actual colour values are the SPA's (mirrored in
//! `web/src/lib/accent.ts`, with tests pinning both sides, like the life-counter layout
//! vocabulary). Free-form colours are deliberately not accepted: every preset ships with
//! WCAG-checked light/dark pairs, which an arbitrary hex could not guarantee.

use crate::error::AppError;

pub const DEFAULT_ACCENT: &str = "pink";

/// Keep in step with `ACCENT_OPTIONS` in `web/src/lib/accent.ts` — a slug added on one
/// side only is either rejected by the API or renders as the default.
pub const SUPPORTED_ACCENTS: &[&str] = &["pink", "ember", "violet", "teal", "blue", "green"];

pub fn is_supported(slug: &str) -> bool {
    SUPPORTED_ACCENTS.contains(&slug)
}

pub fn validate(slug: &str) -> Result<&str, AppError> {
    let slug = slug.trim();
    if is_supported(slug) {
        Ok(slug)
    } else {
        Err(AppError::Validation(format!(
            "accent must be one of {}",
            SUPPORTED_ACCENTS.join(", ")
        )))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn validates_accent_slugs_exactly() {
        // Pins the wire vocabulary the SPA mirrors in web/src/lib/accent.ts.
        assert_eq!(
            SUPPORTED_ACCENTS,
            &["pink", "ember", "violet", "teal", "blue", "green"]
        );
        assert!(SUPPORTED_ACCENTS.contains(&DEFAULT_ACCENT));
        assert_eq!(validate(" teal ").unwrap(), "teal");
        // Exact, case-sensitive membership — no normalisation beyond trimming.
        assert!(validate("Teal").is_err());
        assert!(validate("#ff00aa").is_err());
        assert!(validate("").is_err());
    }
}
